### PR TITLE
Feature/wrapper

### PR DESCRIPTION
**Pull request by jseremba**
Here is a simple tagging component with the following features

- Autocomplete based on a suggestion list.
- Keyboard friendly and mouse support
- Reorder tags using drag and drop
An array of tags that are displayed as pre-selected. Each tag should have an id and a text property which is used to display.

An array of suggestions that are used as basis for showing suggestions

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
+"use strict";
```
Comment (Andries-Smit) [inline Gruntfile.js]: Update grunt file to latest standard. see https://github.com/FlockOfBirds/image-viewer/blob/master/Gruntfile.js

File: karma.conf.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+const webpackConfig = require("./webpack.config")[0];
+const path = require("path");
+Object.assign(webpackConfig, {
```
Comment (Andries-Smit) [inline karma.conf.js]: use webpack merge

File: karma.conf.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        "react/lib/ExecutionEnvironment",
+        "react/lib/ReactContext",
+        "react/addons",
+        "jsdom"
+    ]
+});
+
+module.exports = function(config) {
+    if (config.codeCoverage) {
+        Object.assign(webpackConfig, {
```
Comment (Andries-Smit) [inline karma.conf.js]: use webpack merge

File: package.json [hunk 4]
```diff
@@ -0,0 +1,86 @@
+{
+  "name": "tag",
+  "widgetName": "Tag",
+  "version": "1.0.0",
+  "description": "Tags is a simple tagging component ready to drop in your React projects.",
```
Comment (Andries-Smit) [inline package.json]: React is an implementation detail, description should be functional, from user perspective. Wat can this thing do?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+  "scripts": {
+    "pretest": "tsc",
+    "test": "karma start --single-run --codeCoverage && node tests/remap.js",
+    "emit": "tsc",
+    "lint": "tslint -c tslint.json '**/*.ts' '**/*.tsx' --exclude '**/node_modules/**'"
+  },
+  "pre-commit": [
```
Comment (Andries-Smit) [inline package.json]: Remove pre commit test, they tak to long. move to pre-push
https://github.com/FlockOfBirds/carousel/blob/master/package.json#L15
you probably need some additional package too

File: src/Tag.xml [hunk 6]
```diff
@@ -0,0 +1,87 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<widget id="com.mendix.widget.custom.tag.Tag" needsEntityContext="true" xmlns="http://www.mendix.com/widget/1.0/">
+    <name>Tags</name>
+    <description>Tags is a simple tagging component ready to drop in your React projects.</description>
```
Comment (Andries-Smit) [inline src/Tag.xml]: No periods in any of the descriptions in this file unless you have multiple sentences (which i would not recomment)

File: src/Tag.xml [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    <properties>
+        <property key="afterCreateMicroflow" type="microflow" required="false">
+            <caption>After create tag</caption>
+            <category>Interaction</category>
+            <description>This microflow is triggered after a Tag object is created, committed and added to the reference set.</description>
+            <returnType type="Void"/>
+        </property>
+        <property key="autocompleteMinLength" type="integer" defaultValue='2'>
+            <caption>Autocomplete Min-length</caption>
```
Comment (Andries-Smit) [inline src/Tag.xml]: Only first letter of a caption should be a cappital.
No need for -
Dont use acronyms

File: src/components/Tag.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    minQueryLength: number;
+}
+
+interface TagState {
+    tags: TagItem[];
+    suggestions: string[];
+}
+
+const DefaultClassNames = {
+    remove: "ReactTags__remove",
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: our CSS class name convention is all lower cases and - for separating words.
Do you need theses for the lib? or could you use better names?

File: src/components/Tag.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    suggestions: "ReactTags__suggestions",
+    tagInput: "ReactTags__tagInput",
+    tagInputField: "ReactTags__tagInputField",
+    tags: "ReactTags__tags"
+};
+
+class Tag extends Component<TagProps, TagState> {
+    constructor(props: TagProps) {
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: add newline before constructor

File: src/components/Tag.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        this.setState({ tags });
+    }
+
+    private sortTags() {
+        const tags = this.state.tags
+        if (this.props.sortOrder == "Ascending" && tags.length > 1) {
+            this.setState({
+                tags: tags.sort()
+            });
+        } else {
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: Will fail if length = 0

File: src/components/Tag.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+            this.setState({
+                tags: tags.reverse()
+            });
+        }
+    }
+
+    private handleAddition(tag: string) {
+        const tags = this.state.tags;
+        tags.push({
+            id: tags.length + 1,
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: Why use index as ID. why not use Mendix GUID?

File: src/components/Tag.ts [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+            });
+        }
+        this.setState({ tags });
+    }
+
+    private sortTags() {
+        const tags = this.state.tags
+        if (this.props.sortOrder == "Ascending" && tags.length > 1) {
+            this.setState({
+                tags: tags.sort()
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: Tags is an array of object. Sort will only work on simple arrays.
https://stackoverflow.com/questions/1129216/sort-array-of-objects-by-string-property-value-in-javascript

File: src/components/Tag.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+        tags.splice(position, 1);
+        this.setState({ tags });
+    }
+
+    private handleDrag(tag: TagItem, currPos: number, newPos: number) {
+        const tags = this.state.tags;
+
+        // mutate array
+        tags.splice(currPos, 1);
+        tags.splice(newPos, 0, tag);
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: Should you add a order attribute to your tag object and store the results?

File: src/components/Tag.ts [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+            const tagText = tagElement[index];
+            tags.push({
+                id: tags.length + 1,
+                text: tagText
+            });
+        }
+        this.setState({ tags });
+    }
+
+    private sortTags() {
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: Why would your sort by default, there is an order that can be stored.

File: src/components/Tag.ts [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+        this.setState({ tags });
+    }
+
+    private handleDelete(position: number) {
+        const tags = this.state.tags;
+        tags.splice(position, 1);
+        this.setState({ tags });
+    }
+
+    private handleDrag(tag: TagItem, currPos: number, newPos: number) {
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: Drag drop should be an option, not enabled by default, could conflict with your sort behaviour.
Does this drag and drop work on mobile?

File: src/components/TagContainer.ts [hunk 16]
```diff
@@ -0,0 +1,205 @@
+import { Component, createElement } from "react";
+
+import { Tag } from "./Tag";
+import { SuggestionsData, TagData } from "react-tag-input";
+
+interface WrapperProps {
+    class?: string;
+    mxObject?: mendix.lib.MxObject;
+    style?: string;
+    readOnly: boolean;
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Add implementation for read only see sample of 
https://github.com/FlockOfBirds/boolean-slider/blob/master/src/components/SwitchContainer.ts#L87
Keep in mind you need write right on association and you need to check write right on creation of the object too.

File: src/components/TagContainer.ts [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+
+export interface TagContainerProps extends WrapperProps {
+    afterCreateMicroflow: string;
+    autocompleteMinLength: number;
+    onChangeMicroflow: string;
+    readOnly: boolean;
+    sortOrder: "Ascending" | "Descending";
+    suggestionAttribute: string;
+    suggestionEntity: string;
+    suggestionConstraint: string;
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Should have prop for "Allow create new tags"

File: src/components/TagContainer.ts [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+        this.subscriptionHandles.forEach(mx.data.unsubscribe);
+
+        if (mxObject) {
+            this.subscriptionHandles.push(mx.data.subscribe({
+                callback: this.handleSubscriptions,
+                guid: mxObject.getGuid()
+            }));
+
+            this.subscriptionHandles.push(mx.data.subscribe({
+                attr: this.props.tagAttribute,
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Dont think your context object has this property.

Comment (Andries-Smit) [inline src/components/TagContainer.ts]: You might want to subscribe to association changes... not sure how?
Or to the suggestions ... or tags

File: src/components/TagContainer.ts [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+                guid: mxObject.getGuid()
+            }));
+        }
+    }
+
+    private fetchSuggestions(mxObject: mendix.lib.MxObject) {
+        const xpath = "//" + this.props.suggestionEntity + this.props.suggestionConstraint.replace(/\[\%CurrentObject\%\]/gi, mxObject.getGuid());
+        mx.data.get({
+            xpath: xpath,
+            callback: (objects) => {
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: No parentheses needed around `objects`

File: src/components/TagContainer.ts [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+                const mySuggestion: SuggestionsData[] = objects.map((objects) => ({
+                    text: objects.get(this.props.suggestionAttribute) as string
+                }));
+                const suggestionObjects = mySuggestion.map(value => value.text);
+                this.setState({
+                    suggestions: suggestionObjects
+                });
+            },
+            error: ((error) => {
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: No parentheses needed around `error`

File: src/components/TagContainer.ts [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+                    mxObject.set(this.props.tagAttribute, value);
+                    this.saveTagObject(mxObject);
+                    if (this.props.afterCreateMicroflow) {
+                        this.handleAfterTagCreate();
+                    } else {
+                        console.log(" - please add an after create mf to commit the object, otherwise ui is incorrectly displayed.");
+                    }
+                },
+                error: (Error) => {
+                    console.log("Error creating object: " + Error);
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Error message should tell a bit more, What entity is failing?
Error is an object, so no sens to concatenate them. `error.message`

Not nice to log an error only to the console. The like to know that something went wrong. show a `window.mx.ui.error("message");` for user interaction errors.
Use the `Alert` component for showing initial errors.

This feedback count for all error handlers.

File: src/components/TagContainer.ts [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+                callback: (mxObject) => {
+                    mxObject.set(this.props.tagAttribute, value);
+                    this.saveTagObject(mxObject);
+                    if (this.props.afterCreateMicroflow) {
+                        this.handleAfterTagCreate();
+                    } else {
+                        console.log(" - please add an after create mf to commit the object, otherwise ui is incorrectly displayed.");
+                    }
+                },
+                error: (Error) => {
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Only Classes and Interface should start with capitals, not variable

File: src/components/TagContainer.ts [hunk 23]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    private executeMicroflow(actionName: string, guid: string) {
+        if (actionName) {
+            mx.ui.action(actionName, {
+                error: error =>
+                    mx.ui.error(`Error while executing microflow: ${actionName}: ${error.message}`),
+                params: {
+                    applyto: "selection",
+                    guids: [guid]
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Make sure your linter is on with the latest definition. This line should give you an error.
`[ guid ]` all array and objects with content should be spaced.

File: src/components/TagContainer.ts [hunk 24]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    private handleAfterTagCreate() {
+        const { mxObject, afterCreateMicroflow } = this.props;
+        if (!mxObject || !mxObject.getGuid()) {
+            return;
+        }
+
+        const context = new mendix.lib.MxContext();
+        context.setContext(mxObject.getEntity(), mxObject.getGuid());
```
Comment (Andries-Smit) [inline src/components/TagContainer.ts]: Why create a context if you dont use them? Dont leave any dead code in your commits.

File: src/package.xml [hunk 25]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<package xmlns="http://www.mendix.com/package/1.0/">
+    <clientModule name="Tag" version="1.1.0" xmlns="http://www.mendix.com/clientModule/1.0/">
```
Comment (Andries-Smit) [inline src/package.xml]: Version 1.1.0? Did you already release a public version?

File: src/ui/Tag.css [hunk 26]
```diff
@@ -0,0 +1,62 @@
+/* Example Styles for React Tags*/
+div.ReactTags__tags {
```
Comment (Andries-Smit) [inline src/ui/Tag.css]: No need to set the element before a class this is selective enough `.ReactTags__tags`

File: src/ui/Tag.css [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+    position: relative;
+}
+
+/* Styles for the input */
+div.ReactTags__tagInput {
```
Comment (Andries-Smit) [inline src/ui/Tag.css]: Please dont use `__` in your name. `.class-name-cool`

File: src/ui/Tag.css [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+    border-radius: 2px;
+}
+
+div.ReactTags__selected a.ReactTags__remove {
+    color: #aaa;
+    margin-left: 5px;
+    cursor: pointer;
+}
+
+/* Styles for suggestions */
```
Comment (Andries-Smit) [inline src/ui/Tag.css]: Is this style copied from lib? you can also import styles

File: src/components/Tag.ts [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+        this.handleDelete = this.handleDelete.bind(this);
+        this.handleAddition = this.handleAddition.bind(this);
+        this.handleDrag = this.handleDrag.bind(this);
+        this.sortTags = this.sortTags.bind(this);
+    }
+
+    render() {
+        return DOM.div(
+            {
+                className: classNames(DefaultClassNames.tags)
```
Comment (Andries-Smit) [inline src/components/Tag.ts]: We should use a base class for all our widget starting with `widget-` + name of the widget
=> `widget-tags`

File: webpack.config.js [hunk 30]
```diff
@@ -0,0 +1,45 @@
+const webpack = require("webpack");
+const path = require("path");
+const CopyWebpackPlugin = require("copy-webpack-plugin");
+const ExtractTextPlugin = require("extract-text-webpack-plugin");
+
+const widgetConfig = {
+    entry: "./src/components/TagContainer.ts",
+    output: {
+        path: path.resolve(__dirname, "dist/tmp"),
+        filename: "src/com/mendix/widget/custom/tag/Tag.js",
```
Comment (Andries-Smit) [inline webpack.config.js]: Dont use names in webpack name import the from the packages.json see
https://github.com/FlockOfBirds/image-viewer/blob/master/webpack.config.js

File: src/Tag.xml (hunk 6 above)
Comment (jseremba) [inline src/Tag.xml]: Done

File: src/Tag.xml (hunk 7 above)
Comment (jseremba) [inline src/Tag.xml]: Done

File: src/components/Tag.ts (hunk 8 above)
Comment (jseremba) [inline src/components/Tag.ts]: Done

File: src/components/Tag.ts (hunk 15 above)
Comment (jseremba) [inline src/components/Tag.ts]: It does not respond to the touch events

Comment (jseremba) [inline src/components/Tag.ts]: Drag and drop option added

File: src/components/TagContainer.ts (hunk 17 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: Done

File: src/components/TagContainer.ts (hunk 19 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: Done

File: src/components/TagContainer.ts (hunk 20 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: Done

File: src/components/TagContainer.ts (hunk 22 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: Done

File: src/package.xml (hunk 25 above)
Comment (jseremba) [inline src/package.xml]: Changed version number to 1.0.0

File: src/ui/Tag.css (hunk 28 above)
Comment (jseremba) [inline src/ui/Tag.css]: Yes it is imported from the lib

File: src/ui/Tag.css (hunk 27 above)
Comment (jseremba) [inline src/ui/Tag.css]: Done

File: webpack.config.js (hunk 30 above)
Comment (jseremba) [inline webpack.config.js]: Changed

File: src/ui/Tag.css (hunk 26 above)
Comment (jseremba) [inline src/ui/Tag.css]: Done

File: Gruntfile.js (hunk 1 above)
Comment (jseremba) [inline Gruntfile.js]: Done.

File: karma.conf.js (hunk 2 above)
Comment (jseremba) [inline karma.conf.js]: done

File: karma.conf.js (hunk 3 above)
Comment (jseremba) [inline karma.conf.js]: done

File: package.json (hunk 4 above)
Comment (jseremba) [inline package.json]: changed

Comment (jseremba) [inline package.json]: done

File: package.json (hunk 5 above)
Comment (jseremba) [inline package.json]: done

File: src/components/Tag.ts (hunk 8 above)
Comment (jseremba) [inline src/components/Tag.ts]: Naming changed...
Done

File: src/components/Tag.ts (hunk 29 above)
Comment (jseremba) [inline src/components/Tag.ts]: Done

File: src/components/TagContainer.ts (hunk 23 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: Updated linter with the latest definition

File: src/components/TagContainer.ts (hunk 24 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: Dead code removed

File: src/components/TagContainer.ts (hunk 16 above)
Comment (jseremba) [inline src/components/TagContainer.ts]: readOnly implementation Added

**Comment by jseremba** [comment]
..

**Pull request closed without merging**